File: packages/oxlint-plugin-react-doctor/src/plugin/rules/r3f/r3f-valid-perspective-camera.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { getAuthoritativeJsxAttribute } from "../../utils/get-authoritative-jsx-attribute.js";
import { getStaticObjectPropertyValue } from "../../utils/get-static-object-property-value.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import type { RuleContext } from "../../utils/rule-context.js";
import {
  PERSPECTIVE_CAMERA_ASPECT_ARGUMENT_INDEX,
  PERSPECTIVE_CAMERA_FAR_ARGUMENT_INDEX,
  PERSPECTIVE_CAMERA_NEAR_ARGUMENT_INDEX,
} from "./constants.js";
import { getInvalidPerspectiveCameraParameter } from "./utils/get-invalid-perspective-camera-parameter.js";
import { getStaticNumber } from "./utils/get-static-number.js";
import { hasR3fRuntimeImport } from "./utils/has-r3f-runtime-import.js";
import { isR3fCanvas } from "./utils/is-r3f-canvas.js";

interface StaticR3fCameraParameter {
  node: EsTreeNode;
  value: number;
}

interface StaticR3fCameraParameters {
  aspect: StaticR3fCameraParameter | null;
  far: StaticR3fCameraParameter | null;
  near: StaticR3fCameraParameter | null;
}

const getStaticParameter = (
  expression: EsTreeNode | null | undefined,
  context: RuleContext,
): StaticR3fCameraParameter | null => {
  if (!expression || isNodeOfType(expression, "SpreadElement")) return null;
  const value = getStaticNumber(expression, context.scopes);
  return value === null ? null : { node: expression, value };
};

const getJsxAttributeExpression = (
  node: EsTreeNodeOfType<"JSXOpeningElement">,
  attributeName: string,
): EsTreeNode | null | undefined => {
  const attribute = getAuthoritativeJsxAttribute(node.attributes, attributeName);
  if (!attribute) return undefined;
  if (
    !attribute.value ||
    !isNodeOfType(attribute.value, "JSXExpressionContainer") ||
    isNodeOfType(attribute.value.expression, "JSXEmptyExpression")
  ) {
    return null;
  }
  return attribute.value.expression;
};

const getPerspectiveCameraParameters = (
  node: EsTreeNodeOfType<"JSXOpeningElement">,
  context: RuleContext,
): StaticR3fCameraParameters | null => {
  if (
    !isNodeOfType(node.name, "JSXIdentifier") ||
    node.name.name !== "perspectiveCamera" ||
    node.attributes.some((attribute) => isNodeOfType(attribute, "JSXSpreadAttribute"))
  ) {
    return null;
  }
  const argsExpression = getJsxAttributeExpression(node, "args");
  const args =
    argsExpression && isNodeOfType(argsExpression, "ArrayExpression")
      ? argsExpression.elements
      : [];
  const explicitAspect = getJsxAttributeExpression(node, "aspect");
  const explicitNear = getJsxAttributeExpression(node, "near");
  const explicitFar = getJsxAttributeExpression(node, "far");
  return {
    aspect: getStaticParameter(
      explicitAspect === undefined
        ? args[PERSPECTIVE_CAMERA_ASPECT_ARGUMENT_INDEX]
        : explicitAspect,
      context,
    ),
    near: getStaticParameter(
      explicitNear === undefined ? args[PERSPECTIVE_CAMERA_NEAR_ARGUMENT_INDEX] : explicitNear,
      context,
    ),
    far: getStaticParameter(
      explicitFar === undefined ? args[PERSPECTIVE_CAMERA_FAR_ARGUMENT_INDEX] : explicitFar,
      context,
    ),
  };
};

const getCanvasCameraParameters = (
  node: EsTreeNodeOfType<"JSXOpeningElement">,
  context: RuleContext,
): StaticR3fCameraParameters | null => {
  if (
    !isR3fCanvas(node, context) ||
    getAuthoritativeJsxAttribute(node.attributes, "orthographic")
  ) {
    return null;
  }
  const cameraExpression = getJsxAttributeExpression(node, "camera");
  if (!cameraExpression || !isNodeOfType(cameraExpression, "ObjectExpression")) return null;
  return {
    aspect: getStaticParameter(getStaticObjectPropertyValue(cameraExpression, "aspect"), context),
    near: getStaticParameter(getStaticObjectPropertyValue(cameraExpression, "near"), context),
    far: getStaticParameter(getStaticObjectPropertyValue(cameraExpression, "far"), context),
  };
};

export const r3fValidPerspectiveCamera = defineRule({
  id: "r3f-valid-perspective-camera",
  title: "Invalid R3F perspective camera",
  category: "Correctness",
  tags: ["react-jsx-only"],
  severity: "error",
  recommendation:
    "Use a positive aspect ratio and near plane, and keep the far plane greater than the near plane",
  create: (context: RuleContext) => {
    let importsReactThreeFiber = false;
    return {
      Program(node: EsTreeNodeOfType<"Program">) {
        importsReactThreeFiber = hasR3fRuntimeImport(node, context.scopes);
      },
      JSXOpeningElement(node: EsTreeNodeOfType<"JSXOpeningElement">) {
        if (!importsReactThreeFiber) return;
        const parameters =
          getPerspectiveCameraParameters(node, context) ?? getCanvasCameraParameters(node, context);
        if (!parameters) return;
        const invalidParameter = getInvalidPerspectiveCameraParameter(parameters);
        if (!invalidParameter) return;
        context.report({
          node: invalidParameter.node,
          message: invalidParameter.message,
        });
      },
    };
  },
});

File: native/oxlint/rules/r3f-valid-perspective-camera.rs
use oxc_ast::AstKind;
use oxc_diagnostics::OxcDiagnostic;
use oxc_span::GetSpan;

use crate::{
    context::{ContextHost, LintContext},
    rule::{Rule, RuleCategory, RuleInfo, RuleMeta},
};

const ASPECT_ARGUMENT_INDEX: usize = 1;
const NEAR_ARGUMENT_INDEX: usize = 2;
const FAR_ARGUMENT_INDEX: usize = 3;

#[derive(Debug, Default, Clone)]
pub struct R3FValidPerspectiveCamera;

impl RuleMeta for R3FValidPerspectiveCamera {
    const NAME: &'static str = "r3f-valid-perspective-camera";
    const PLUGIN: &'static str = "react_doctor_native";
    const CATEGORY: RuleCategory = RuleCategory::Correctness;
    const VERSION: &'static str = "0.1.0";
    const INFO: RuleInfo = RuleInfo {
        short_description: "Validate React Three Fiber perspective cameras.",
    };
}

impl Rule for R3FValidPerspectiveCamera {
    fn should_run(&self, ctx: &ContextHost) -> bool {
        ctx.source_type().is_jsx()
    }

    fn run_once<'a>(&self, ctx: &LintContext<'a>) {
        if file_is_non_react_jsx_dialect(ctx) || !has_r3f_runtime_import(ctx) {
            return;
        }
        for node in ctx.nodes().iter() {
            let AstKind::JSXOpeningElement(opening_element) = node.kind() else {
                continue;
            };
            let parameters = match &opening_element.name {
                oxc_ast::ast::JSXElementName::Identifier(identifier)
                    if identifier.name == "perspectiveCamera"
                        && !opening_element.attributes.iter().any(|attribute| {
                            matches!(
                                attribute,
                                oxc_ast::ast::JSXAttributeItem::SpreadAttribute(_)
                            )
                        }) =>
                {
                    let arguments = get_authoritative_jsx_attribute(
                        opening_element,
                        "args",
                        true,
                    )
                    .and_then(|attribute| jsx_attribute_expression(attribute))
                    .and_then(|expression| {
                        let oxc_ast::ast::Expression::ArrayExpression(arguments) =
                            expression.get_inner_expression()
                        else {
                            return None;
                        };
                        Some(arguments)
                    });
                    let parameter_expression = |property_name, argument_index| {
                        get_authoritative_jsx_attribute(opening_element, property_name, true)
                            .map_or_else(
                                || {
                                    arguments
                                        .and_then(|arguments| {
                                            arguments.elements.get(argument_index)
                                        })
                                        .and_then(
                                            oxc_ast::ast::ArrayExpressionElement::as_expression,
                                        )
                                },
                                |attribute| jsx_attribute_expression(attribute),
                            )
                    };
                    Some((
                        parameter_expression("aspect", ASPECT_ARGUMENT_INDEX),
                        parameter_expression("near", NEAR_ARGUMENT_INDEX),
                        parameter_expression("far", FAR_ARGUMENT_INDEX),
                    ))
                }
                _ if is_r3f_canvas(opening_element, ctx)
                    && get_authoritative_jsx_attribute(opening_element, "orthographic", true)
                        .is_none() =>
                {
                    let Some(camera_expression) = get_authoritative_jsx_attribute(
                        opening_element,
                        "camera",
                        true,
                    )
                    .and_then(|attribute| jsx_attribute_expression(attribute))
                    else {
                        continue;
                    };
                    Some((
                        get_static_object_property_value(camera_expression, "aspect"),
                        get_static_object_property_value(camera_expression, "near"),
                        get_static_object_property_value(camera_expression, "far"),
                    ))
                }
                _ => None,
            };
            let Some((aspect_expression, near_expression, far_expression)) = parameters else {
                continue;
            };
            let aspect = resolve_camera_parameter(aspect_expression, ctx);
            let near = resolve_camera_parameter(near_expression, ctx);
            let far = resolve_camera_parameter(far_expression, ctx);
            let invalid_parameter = if let Some((expression, value)) = aspect
                && value <= 0.0
            {
                Some((
                    expression,
                    "This perspective camera has a non-positive aspect ratio, so its projection is invalid",
                ))
            } else if let Some((expression, value)) = near
                && value <= 0.0
            {
                Some((
                    expression,
                    "This perspective camera has a non-positive near plane, but Three.js requires near to be greater than zero",
                ))
            } else if let Some((expression, value)) = far
                && value <= 0.0
            {
                Some((
                    expression,
                    "This perspective camera has a non-positive far plane, but Three.js requires far to be greater than its positive near plane",
                ))
            } else if let (Some((_, near_value)), Some((far_expression, far_value))) = (near, far)
                && far_value <= near_value
            {
                Some((
                    far_expression,
                    "This perspective camera's far plane is not greater than its near plane, so its projection is invalid",
                ))
            } else {
                None
            };
            if let Some((expression, message)) = invalid_parameter {
                ctx.diagnostic(OxcDiagnostic::error(message).with_label(expression.span()));
            }
        }
    }
}

fn resolve_camera_parameter<'a>(
    expression: Option<&'a oxc_ast::ast::Expression<'a>>,
    ctx: &LintContext<'a>,
) -> Option<(&'a oxc_ast::ast::Expression<'a>, f64)> {
    expression.and_then(|expression| {
        resolve_static_number(expression, ctx).map(|value| (expression, value))
    })
}
